crates/adapters/src/transport/kafka/mod.rs

```diff
@@ -359,3 +361,61 @@ fn generate_oauthbearer_token(
         lifetime_ms: i64::MAX,
     })
 }
+
+/// Tracks and reports memory use for a consumer or producer.
+struct MemoryUseReporter {
+    /// When we were created.
+    ///
+    /// We don't want to report on memory use for a while afterward, since it
+    /// will take some time to reach what we hope is a steady state.
+    start: Instant,
+
+    last_report: Option<(Instant, u64)>,
+}
+
+impl MemoryUseReporter {
+    fn new() -> Self {
+        Self {
+            start: Instant::now(),
+            last_report: None,
+        }
+    }
+    fn update(&mut self, statistics: &Statistics) {
+        /// Minimum time before first report.
+        const REPORT_DELAY: Duration = Duration::from_secs(60);
+
+        /// Minimum amount of memory to report on.
+        const MIN_MEMORY: u64 = 1024 * 1024;
+
+        if self.start.elapsed() < REPORT_DELAY {
+            return;
+        }
+
+        let mut memory = 0;
+        for topic in statistics.topics.values() {
+            for partition in topic.partitions.values() {
+                memory += partition.msgq_bytes + partition.xmit_msgq_bytes + partition.fetchq_size;
+            }
+        }
+        match &self.last_report {
+            None if memory > MIN_MEMORY => {
+                info!(
+                    "Buffered {} after {} seconds",
+                    HumanBytes::new(memory),
+                    self.start.elapsed().as_secs()
+                );
+            }
+            Some((last_time, last_memory)) if memory > *last_memory * 3 / 2 => {
+                info!(
+                    "Buffers grew {:.0}%, from {} to {}, in last {} seconds",
```

Comment:
Could the break down of the memory usage by partition and by counter be helpful?

Reply:
I doubt it.

For partitions, I don't have any reason to suspect that distinguishing partitions is useful.

There are three counters. Consumers use only one and producers use the other two. The distinction between the two that the producers use is some kind of undocumented internal detail.

I think that we'll probably just use this to find out whether librdkafka is really the culprit or not.